# Student Performance Indicator

### Life Cycle of a Machine Learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to Perform
- Exploratory Data Analysis (EDA)
- Data Pre-Processing
- Model Training
- Choose Best Model

### 1) Problem Statement

This project aims to understand and predict student performances (exam scores) using known details about students such as **Gender**, **Ethnicity**, **Parental level of education**, **Lunch** and **Test Preparation Course**.

### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams

- The dataset contains 8 columns and 1000 rows


### 2.1) Import Data and Required Packages

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
%matplotlib inline
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('data/stud.csv')

In [4]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [6]:
# Check the shape of the data
print(f"Number of rows: {df.shape[0]}\nNumber of columns: {df.shape[1]}")

Number of rows: 1000
Number of columns: 8


In [9]:
df["test_preparation_course"].unique()

array(['none', 'completed'], dtype=object)

### 2.2) Dateset Information

- **gender**: Student gender -> (Male/Female)
- **race_ethnicity**: Race/Ethnicity of Student -> (Group A,B,C,D,E)
- **parental_level_of_education**: Highest level of education achieved by the student's parents -> (bachelor's degree, some college, master's degree,
       associate's degree, high school, some high school)

- **lunch**: What lunch did the student have before the test -> (standard or free/reduced)
- **test_preparation_course**: Did the student complete a test preparation course prior to the test? -> (none/completed)
- **math score**
- **reading score**
- **writing score**

### 3) Data Checks to Perform

- Check for missing data
- Check for duplicate records
- Check data types
- Check number of unique values for each column
- Check statistics of dataset
- Check various categories present in the categorical columns

#### 3.1) Check for Missing Data

In [17]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

There are no missing values in the dateset

#### 3.2) Check for Duplicate Records

In [27]:
int(df.duplicated().sum())

0

There are no duplicates in the dateset

#### 3.3) Check Data Types

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


#### 3.4) Checking the Number of Unique Values in Each Column

In [30]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

#### 3.5) Check the Statistics of the Dateset

In [33]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

##### Insights from Stats
- We can see that the average score for math, reading and writing tests are quite close, ranging between ~66.09 and ~69.17
- The standard deviations for the three test scores are also very similar, ranging between ~14.60 and ~15.20
- The mimimum test score is highest for the reading test (17.0) then writing test (10.0) and then math (0.0)
- At least one student from each test achieved full marks (100.0)